src/utils/dom-tranform-stream.js

```diff
@@ -0,0 +1,142 @@
+/**
+ * Copyright 2020 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {Deferred} from './promise';
+import {Services} from '../services';
+import {devAssert, user} from '../log';
+import {removeNoScriptElements} from './dom-writer';
+
+export class DomTransformStream {
+  /**
+   * @param {!Window} win
+   */
+  constructor(win) {
+    const headDefer = new Deferred();
+    /** @const @private {!Promise<!Element>} */
+    this.headPromise_ = headDefer.promise;
+
+    /** @const @private {!function()} */
+    this.headResolver_ = headDefer.resolve;
+
+    const transferDefer = new Deferred();
+    /** @const @private {!Promise} */
+    this.bodyTransferPromise_ = transferDefer.promise;
+
+    /** @const @private {!function()} */
+    this.bodyTransferResolver_ = transferDefer.resolve;
+
+    /** @private {?Element} */
+    this.detachedBody_ = null;
+
+    /** @private {?Promise} */
+    this.currentChunkTransferPromise_ = null;
+
+    /** @private {boolean} */
+    this.shouldTransfer_ = false;
+
+    /** @private {boolean} */
+    this.mergeScheduled_ = false;
+
+    /** @const @private */
+    this.vsync_ = Services.vsyncFor(win);
+  }
+
+  /**
+   * Callback passed into DetachedDomStream constructor. Receives updated
+   * document everytime a new chunk is written.
+   * Resolves headPromise when body is available, and streams to body iff
+   * bodyTransfer() has been called.
+   * @param {!Document} detachedDoc
+   */
+  onChunk(detachedDoc) {
+    // <body> is newly formed.
+    if (!this.detachedBody_ && detachedDoc.body) {
+      this.detachedBody_ = detachedDoc.body;
+      this.headResolver_(detachedDoc.head);
+    }
+
+    // If bodyTransfer has already been called, keep transferring on new chunks.
+    if (this.shouldTransfer_) {
+      this.transferBodyChunk_();
+    }
+  }
+
+  /**
+   * Callback passed into DetachedDomStream constructor. Called with complete
+   * doc when stream is closed.
+   * Schedules final transfer, then resovles body complete promise.
+   * @param {!Document} unusedCompleteDoc
+   */
+  onEnd(unusedCompleteDoc) {
+    this.transferBodyChunk_().then(this.bodyTransferResolver_);
```

Comment:
It'd be cleaner:

```
this.bodyTransferResolver_(this.transferBodyChunk_())
```

For instance, if for some reason `this.transferBodyChunk_()` gets rejected, you will catch this case. With the `this.transferBodyChunk_().then(...)` you miss out on rejection.

Reply:
nice, done.